File: frontend/src/features/production/api/procedurePrices.ts
import { service } from '@/api/request'
import type {
  ProcedurePriceListItem,
  ProcedurePriceScope,
  ProcedurePriceRevision,
} from '../domain/procedurePrices'

export async function queryProcedurePrices(
  departmentCode: string, page: number, pageSize: number, keyword: string,
) {
  const response = await service.get<{ data: ProcedurePriceListItem[]; total: number }>(
    `/departments/${departmentCode}/procedure-prices`,
    { params: { page, page_size: pageSize, keyword: keyword.trim() || undefined } },
  )
  return response.data
}

export async function queryProcedurePriceRevisions(
  departmentCode: string,
  page = 1,
  pageSize = 50,
) {
  const response = await service.get<{ data: ProcedurePriceRevision[]; total: number }>(
    `/departments/${departmentCode}/procedure-price-revisions`,
    { params: { page, page_size: pageSize } },
  )
  return response.data
}

export async function saveProcedurePrices(
  departmentCode: string,
  scope: ProcedurePriceScope,
  procedures: Array<{
    procedure_id: number | null
    procedure_name: string
    unit_price: number | null
  }>,
) {
  await service.put(
    `/departments/${departmentCode}/procedure-prices/${scope.product_id}/${scope.product_version}/${encodeURIComponent(scope.origin_flow_node_id)}/${encodeURIComponent(scope.flow_node_id)}`,
    { procedures },
  )
}

export async function confirmProcedurePrices(
  departmentCode: string,
  scope: ProcedurePriceScope,
  procedures: Array<{
    procedure_id: number | null
    procedure_name: string
    unit_price: number | null
  }>,
) {
  await service.post(
    `/departments/${departmentCode}/procedure-prices/${scope.product_id}/${scope.product_version}/${encodeURIComponent(scope.origin_flow_node_id)}/${encodeURIComponent(scope.flow_node_id)}/confirm`,
    { procedures },
  )
}

export async function cancelProcedurePriceConfirmation(
  departmentCode: string,
  scope: ProcedurePriceScope,
) {
  await service.delete(
    `/departments/${departmentCode}/procedure-prices/${scope.product_id}/${scope.product_version}/${encodeURIComponent(scope.origin_flow_node_id)}/${encodeURIComponent(scope.flow_node_id)}/confirm`,
  )
}

export async function saveTemporaryWorkOrderPrice(
  departmentCode: string,
  workOrderId: number,
  unitPrice: number | null,
) {
  await service.put(
    `/departments/${departmentCode}/procedure-prices/temporary-work-orders/${workOrderId}`,
    { unit_price: unitPrice },
  )
}

File: backend/src/routers/procedure_prices.py
from fastapi import APIRouter, Depends, Query, Response

from authorization import require_any_permission
from domain.permissions import PRODUCTION_MANAGE, PRODUCTION_VIEW
from modules.standard_execution.api import (
    confirm_procedure_configuration,
    list_procedure_prices,
    update_procedure_price,
)
from schemas.procedure_prices import ProcedurePriceListEnvelope, ProcedurePriceUpdate


router = APIRouter(tags=["procedure-prices"])


@router.get(
    "/departments/{department_code}/procedure-prices",
    response_model=ProcedurePriceListEnvelope,
)
def procedure_prices(
    department_code: str,
    page: int = Query(default=1, gt=0),
    page_size: int = Query(default=20, gt=0, le=100),
    keyword: str | None = Query(default=None, max_length=200),
    user: dict = Depends(require_any_permission(PRODUCTION_VIEW)),
):
    data, total = list_procedure_prices(
        department_code,
        page,
        page_size,
        keyword,
        user["department"],
        user["is_system"],
    )
    return {"data": data, "total": total}


@router.put(
    "/departments/{department_code}/procedure-prices/"
    "{product_id}/{product_version}/{origin_flow_node_id}/{flow_node_id}",
    status_code=204,
)
def procedure_price_update(
    department_code: str,
    product_id: int,
    product_version: int,
    origin_flow_node_id: str,
    flow_node_id: str,
    payload: ProcedurePriceUpdate,
    user: dict = Depends(require_any_permission(PRODUCTION_MANAGE, csrf=True)),
):
    update_procedure_price(
        department_code,
        product_id,
        product_version,
        origin_flow_node_id,
        flow_node_id,
        payload,
        user["department"],
        user["is_system"],
        user["username"],
    )
    return Response(status_code=204)


@router.post(
    "/departments/{department_code}/procedure-prices/"
    "{product_id}/{product_version}/{origin_flow_node_id}/{flow_node_id}/confirm",
    status_code=204,
)
def procedure_configuration_confirm(
    department_code: str,
    product_id: int,
    product_version: int,
    origin_flow_node_id: str,
    flow_node_id: str,
    payload: ProcedurePriceUpdate,
    user: dict = Depends(require_any_permission(PRODUCTION_MANAGE, csrf=True)),
):
    confirm_procedure_configuration(
        department_code,
        product_id,
        product_version,
        origin_flow_node_id,
        flow_node_id,
        payload,
        user["department"],
        user["is_system"],
        user["username"],
    )
    return Response(status_code=204)
